From bd5d699e83e2dacf7bb71878aeaee88c4d164086 Mon Sep 17 00:00:00 2001
From: Hannes Reinecke <hare@suse.de>
Date: Wed, 27 Apr 2016 16:12:05 +0200
Subject: [PATCH] scsi_dh_rdac: retry inquiry for UNIT ATTENTION
References: bsc#934760
Patch-Mainline: submitted linux-scsi 2016/09/28

When we receive a UNIT ATTENTION upon sending any INQUIRY
command we should just retry the command to clear any
non-permanent UNIT ATTENTION.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/device_handler/scsi_dh_rdac.c | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/drivers/scsi/device_handler/scsi_dh_rdac.c b/drivers/scsi/device_handler/scsi_dh_rdac.c
index 06fbd0b..bb2efb7 100644
--- a/drivers/scsi/device_handler/scsi_dh_rdac.c
+++ b/drivers/scsi/device_handler/scsi_dh_rdac.c
@@ -405,12 +405,14 @@ static int submit_inquiry(struct scsi_device *sdev, int page_code,
 {
 	struct request *rq;
 	struct request_queue *q = sdev->request_queue;
-	int err = SCSI_DH_RES_TEMP_UNAVAIL;
+	struct scsi_sense_hdr sense_hdr;
+	int err = SCSI_DH_IO, retry_cnt = RDAC_RETRY_COUNT;
 
 	rq = get_rdac_req(sdev, &h->inq, len, READ);
 	if (!rq)
-		goto done;
+		return SCSI_DH_RES_TEMP_UNAVAIL;
 
+retry:
 	/* Prepare the command. */
 	rq->cmd[0] = INQUIRY;
 	rq->cmd[1] = 1;
@@ -422,12 +424,21 @@ static int submit_inquiry(struct scsi_device *sdev, int page_code,
 	memset(rq->sense, 0, SCSI_SENSE_BUFFERSIZE);
 	rq->sense_len = 0;
 
-	err = blk_execute_rq(q, NULL, rq, 1);
-	if (err == -EIO)
-		err = SCSI_DH_IO;
+	if (blk_execute_rq(q, NULL, rq, 1) == -EIO) {
+		if (scsi_normalize_sense(rq->sense, SCSI_SENSE_BUFFERSIZE,
+					 &sense_hdr)) {
+			if (sense_hdr.sense_key == UNIT_ATTENTION) {
+				if (retry_cnt) {
+					retry_cnt--;
+					goto retry;
+				}
+				err = SCSI_DH_RETRY;
+			}
+		}
+	} else
+		err = SCSI_DH_OK;
 
 	blk_put_request(rq);
-done:
 	return err;
 }
 
-- 
1.8.5.6

